cpp/bench/sg/dataset.cuh

```diff
@@ -0,0 +1,164 @@
+ * @tparam D type of the dataset (type of X)
+ * @tparam L type of the labels/output (type of y)
+ */
+template <typename D, typename L>
+struct Dataset {
+  /** input data */
+  D* X;
+  /** labels or output associated with each row of input data */
+  L* y;
+
+  /** allocate space needed for the dataset */
+  void allocate(const cumlHandle& handle, const DatasetParams& p) {
+    auto allocator = handle.getDeviceAllocator();
+    auto stream = handle.getStream();
+    X = (D*)allocator->allocate(p.nrows * p.ncols * sizeof(D), stream);
+    y = (L*)allocator->allocate(p.nrows * sizeof(L), stream);
+  }
+
+  /** free-up the buffers */
+  void deallocate(const cumlHandle& handle, const DatasetParams& p) {
+    auto allocator = handle.getDeviceAllocator();
+    auto stream = handle.getStream();
+    allocator->deallocate(X, p.nrows * p.ncols * sizeof(D), stream);
+    allocator->deallocate(y, p.nrows * sizeof(L), stream);
+  }
+
+  /** whether the current dataset is for classification or regression */
+  bool isClassification() const { return typeid(D) != typeid(L); }
+
+  /**
+   * Generate random blobs data. Args are the same as in make_blobs.
+   * Assumes that the user has already called `allocate`
+   */
+  void blobs(const cumlHandle& handle, const DatasetParams& p,
+             const BlobsParams& b) {
+    ASSERT(isClassification(),
+           "make_blobs: is only for classification/clustering problems!");
+    auto* tmpX = X;
+    auto allocator = handle.getDeviceAllocator();
+    auto stream = handle.getStream();
+    if (!p.rowMajor) {
+      tmpX = (D*)allocator->allocate(p.nrows * p.ncols * sizeof(D), stream);
+    }
+    MLCommon::Random::make_blobs<D, int>(
+      tmpX, y, p.nrows, p.ncols, p.nclasses, allocator, stream, nullptr,
+      nullptr, D(b.cluster_std), b.shuffle, D(b.center_box_min),
+      D(b.center_box_max), b.seed);
+    if (!p.rowMajor) {
+      D alpha = (D)1.0, beta = (D)0.0;
+      MLCommon::LinAlg::cublasgeam<D>(
+        handle.getImpl().getCublasHandle(), CUBLAS_OP_T, CUBLAS_OP_N, p.nrows,
+        p.ncols, &alpha, tmpX, p.ncols, &beta, X, p.nrows, X, p.nrows, stream);
+      allocator->deallocate(tmpX, p.nrows * p.ncols * sizeof(D), stream);
+    }
+  }
+
+  /**
+   * @brief Read the input csv file and construct the dataset.
+   *        Assumes that the user has already called `allocate`
+   * @tparam Lambda lambda to customize how to read the data and labels
+   * @param handle cuml handle
+   * @param csvfile the csv file
+   * @param p dataset parameters
+   * @param readOp functor/operator to take the current row of values and update
+   *               the dataset accordingly. Its signature is expected to be:
+   * `void readOp(const std::vector<std::string>& row, std::vector<D>& X,
+   *              std::vector<L>& y, int lineNum, const DatasetParams& p);`
+   */
+  template <typename Lambda>
+  void read_csv(const cumlHandle& handle, const std::string& csvfile,
```

Comment:
I would still love to see this pull out into a common area in the codebase (along with some differen CSV files) so that we can use it for our gtests as well.

Reply:
Agree with you, in general. I have filed issue #1155 for this.

Reply:
Let's discuss more details about this in the above issue.